### PR TITLE
add automatic persisted queries implementation to client

### PR DESCRIPTION
**Pull request by gumimin**
### :pencil: Description
- add apq interfaces to client
- add apq implementation to ktor client
- add apq implementation to spring client

### :link: Related Issues
https://github.com/ExpediaGroup/graphql-kotlin/issues/1640

### CODE REVIEW

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesExtension.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.client.types
+
+interface AutomaticPersistedQueriesExtension {
```
Comment (dariuszkuc) [inline clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesExtension.kt:19]: why is this an interface and not just a data class?

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.client.types
+
+interface AutomaticPersistedQueriesSettings {
```
Comment (dariuszkuc) [inline clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: why not data class?

File: clients/graphql-kotlin-ktor-client/src/main/kotlin/com/expediagroup/graphql/client/ktor/GraphQLKtorClient.kt [hunk 3]
```diff
@@ -39,16 +48,52 @@ import java.net.URL
...
-            setBody(TextContent(serializer.serialize(request), ContentType.Application.Json))
-        }.body()
-        return serializer.deserialize(rawResult, request.responseType())
+        val queryId = request.getQueryId()
+        val automaticPersistedQueriesExtension = object : AutomaticPersistedQueriesExtension {
+            override val version: Int
+                get() = automaticPersistedQueriesSettings.version
+            override val sha256Hash: String
+                get() = queryId
+        }
```
Comment (dariuszkuc) [inline clients/graphql-kotlin-ktor-client/src/main/kotlin/com/expediagroup/graphql/client/ktor/GraphQLKtorClient.kt:62]: this should only be computed IF using APQ

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesExtension.kt (hunk 1 above)
Comment (gumimin) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesExtension.kt]: There were times during my implementation when it was more convenient to use interface, but now it is no longer necessary to use interface, so I'll change it to data class.

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt (hunk 2 above)
Comment (gumimin) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt]: I'll also change this to data class.

**Comment by gumimin** [comment]
@dariuszkuc 
Thank you for your reviewing!
I've fixed points you commented.

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.client.types
+
+data class AutomaticPersistedQueriesSettings(
```
Comment (samuelAndalon) [inline clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: APQ requests can also be send through HTTP POST, this implementation forces GET, maybe in the APQSetting we can specify which HTTP Verb the client wish to use.

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/extensions/AutomaticPersistedQueriesExtensions.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+fun GraphQLClientRequest<*>.getQueryId(): String =
+    String.format(
+        "%064x",
+        BigInteger(1, MESSAGE_DIGEST.digest(this.query.toByteArray(StandardCharsets.UTF_8)))
+    ).also {
+        MESSAGE_DIGEST.reset()
+    }
+
+fun AutomaticPersistedQueriesExtension.toQueryParamString() = """{"persistedQuery":{"version":$version,"sha256Hash":"$sha256Hash"}}"""
```
Comment (samuelAndalon) [inline clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/extensions/AutomaticPersistedQueriesExtensions.kt:19]: whereas POST can also have query params , when using GraphQL with POST you have to pass query, operationName and extensions in the payload -- i would just call it `serialize()`.

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    val enabled: Boolean,
+    val version: Int
+)
+
+val defaultAutomaticPersistedQueriesSettings = AutomaticPersistedQueriesSettings(
+    enabled = false,
+    version = 1
```
Comment (samuelAndalon) [inline clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:26]: For now i would just expose the `enabled` setting, not the version, other APQ implementations does not allow specifying the version as of now there only exists 1.

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt (hunk 4 above)
Comment (gumimin) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: @samuelAndalon 
I thought APQ had to be GET.
[The Apollo documentation](https://www.apollographql.com/docs/apollo-server/performance/apq/
) was also written in GET.

For a POST request in GraphQL, [the query field in the body should be required](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md).
If APQ allows POST requests, then inevitably a large query string will be included in the request, I think which is inconsistent with the benefits of APQ.

Am I correct that the query field is optional when APQ is used in the POST method?
If so, I will change `interface GraphQLClientRequest.query` to nullable.
Also can you please share any detailed documentation on APQ specs (especially HTTP request and response)?

Comment (samuelAndalon) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: for APQ requests, the `query` field of the JSON payload is not required. 
graphql kotlin server allows APQ requests for both GET and POST, i think we should be consistent with the client.

Comment (samuelAndalon) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: >If APQ allows POST requests, then inevitably a large query string will be included in the request, I think which is inconsistent with the benefits of APQ.

APQ worst case scenario (when persited query is not found) needs to send a second request, with persitedQuery extension and query

Comment (samuelAndalon) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: https://opensource.expediagroup.com/graphql-kotlin/docs/server/automatic-persisted-queries/

Comment (dariuszkuc) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: AFAIK there is no official APQ spec yet and there are number of implementations out there (including the linked one from Apollo). Per the graphql http spec [discussion](https://github.com/graphql/graphql-over-http/issues/216) and [issue](https://github.com/graphql/graphql-over-http/issues/38) it is still TBD.

As for changing the `query` to be nullable -> I have no idea what sort of impact that would be, guessing it should be ok as it is a handled by the client internals.

Comment (samuelAndalon) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: The server logic itself should handle the nullable `query`, at least graphql-kotlin server does that.

if you dont send an APQ request (request without entension), and no query, then the server will return 400.

Comment (samuelAndalon) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: when added APQ in graphql-kotlin had to make query value default to ""

https://github.com/ExpediaGroup/graphql-kotlin/commit/f773327b126832c6176c0fd64e26aa6fea023d94#diff-b4c1162353124139fdbd7ef3c7f304d95203c1c7b8392ed9a153ed7a4c67f96eR42

**Comment by samuelAndalon** [comment]
@gumimin can you take a look at the build failure, i think its coverage, we might need to add a test for GET and POST

**Comment by gumimin** [comment]
@samuelAndalon 
I was aware of this but did not have time to implement. I will add more tests over the weekend, so please wait.

File: clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt (hunk 4 above)
Comment (gumimin) [reply clients/graphql-kotlin-client/src/main/kotlin/com/expediagroup/graphql/client/types/AutomaticPersistedQueriesSettings.kt:19]: @samuelAndalon @dariuszkuc 
Thank you for your comment!
I've change to allow POST requests and `GraphQLClientRequest.query` to be nullable.

**Comment by gumimin** [comment]
@samuelAndalon 
I've fixed the test, please review again.

**Review by samuelAndalon — APPROVED** [review]

**Comment by samuelAndalon** [comment]
@gumimin Thank you for your contribution!

**Pull request merged**